README.md
<div align="center">

[![npm package @latest](https://img.shields.io/npm/v/node-audiorecorder.svg?label=npm@latest&style=flat-square&maxAge=3600)](https://npmjs.com/package/node-audiorecorder)
[![License agreement](https://img.shields.io/github/license/redkenrok/node-audiorecorder.svg?style=flat-square&maxAge=86400)](https://github.com/redkenrok/node-audiorecorder/blob/master/LICENSE)
[![Open issues on GitHub](https://img.shields.io/github/issues/redkenrok/node-audiorecorder.svg?style=flat-square&maxAge=86400)](https://github.com/redkenrok/node-audiorecorder/issues)

</div>

# Audio recorder

Audio recorder for [Node.js](https://nodejs.org/), delivers a 16-bit signed-integer linear pulse modulation WAV stream. Based of [Gilles De Mey](https://github.com/gillesdemey)'s [node-record-lpcm16](https://github.com/gillesdemey/node-record-lpcm16).

## Installation

```
npm install node-audiorecorder
```

## Dependencies

This module requires you to install [SoX](http://sox.sourceforge.net/) and it must be available in your \$PATH.

### For Linux

```
sudo apt-get install sox libsox-fmt-all
```

### For MacOS

```
brew install sox
```

### For Windows

[Download the binaries](http://sourceforge.net/projects/sox/files/latest/download)

## Usage

### Constructor

```javascript
// Import module.
const AudioRecorder = require('node-audiorecorder')

// Options is an optional parameter for the constructor call.
// If an option is not given the default value, as seen below, will be used.
const options = {
  program: `rec`, // Which program to use, either `arecord`, `rec`, or `sox`.
  device: null, // Recording device to use, e.g. `hw:1,0`

  bits: 16, // Sample size. (only for `rec` and `sox`)
  channels: 1, // Channel count.
  encoding: `signed-integer`, // Encoding type. (only for `rec` and `sox`)
  format: `S16_LE`, // Encoding type. (only for `arecord`)
  rate: 16000, // Sample rate.
  type: `wav`, // Format type.

  // Following options only available when using `rec` or `sox`.
  silence: 2, // Duration of silence in seconds before it stops recording.
  thresholdStart: 0.5, // Silence threshold to start recording.
  thresholdStop: 0.5, // Silence threshold to stop recording.
  keepSilence: true, // Keep the silence in the recording.
}
// Optional parameter intended for debugging.
// The object has to implement a log and warn function.
const logger = console

// Create an instance.
let audioRecorder = new AudioRecorder(options, logger)
```

> If you can't capture any sound with 'arecord' try to running 'arecord -l' to which devices are available.

> See the [arecord documentation](https://linux.die.net/man/1/arecord) for more detail on its options.

> See the [sox documentation](http://sox.sourceforge.net/Docs/Documentation) for more detail on the rec and sox options.

### Methods

```javascript
// Creates and starts the recording process.
audioRecorder.start()
// Stops and removes the recording process.
audioRecorder.stop()
// Returns the stream of the recording process.
audioRecorder.stream()
```

### Examples

See the [examples directory](https://github.com/RedKenrok/node-audiorecorder/tree/master/examples) for example usage.

> For another example see the [node-hotworddetector](https://github.com/RedKenrok/node-hotworddetector) module, or [Electron-VoiceInterfaceBoilerplate](https://github.com/RedKenrok/Electron-VoiceInterfaceBoilerplate)'s input.js.

## Troubleshooting

### Windows continues recording

If you have issues with continues recording on Windows 10 with SoX 14.4.2 or later, install version [14.4.1](https://sourceforge.net/projects/sox/files/sox/14.4.1/) instead.

## License

[MIT license](https://github.com/redkenrok/node-audiorecorder/blob/master/LICENSE)


CHANGELOG.md
# Changelog

# 3.0.0 (2022-02-14)

### Changed

- Allow for multiple instance at once.
- Updated development dependencies.

# 2.0.1 (2020-04-28)

### Changed

- Updated development dependencies.

# 2.0.0 (2020-01-16)

### Removed

- Device option removed for `sox` and `rec` commands.

## 1.4.1 (2019-03-13)

### Changed

- Updated dependencies.

## 1.4.0 (2018-12-27)

### Added

- `keepSilence` option added fo `sox` and `rec` only. The option enables or disabled the `-l` flag which controls whether the silence is part of the recording or not.

## 1.3.0-beta.0 (2018-12-12)

### Added

- Opened up the `format` argument of the audio recorder options. (only for `arecord`)

### Changed

- Fixed `arecord` command generation.

## 1.2.0 (2018-11-09)

### Changed

- Added `-l` to silence effect, therefore the silence not removed from the start of the recording.
- Default values of `thresholdStart` and `thresholdStop` are both set to `1`.
- Fixed duration of silence effect since duration parameters need to have a decimal. For example `2.0` instead of `2`.

### Removed

- `sampleRate` and `threshold` options removed.

## 1.1.6 (2018-11-08)

### Added

- Example added to examples directory, and changed `example.js` to `examples/print-command.js`.

### Fixed

- Fixed silence effect.

## 1.1.5 (2018-10-24)

### Added

- Added `bits`, `encoding`, `rate`, and `type` properties to options, which allows greater control over output.

### Changed

- `sampleRate` option renamed to `rate`, legacy support still available.

## 1.1.4 (2018-10-23)

### Added

- Tests added with continues integration and code coverage.

### Changed

- Clarified examples in `README.md`.

### Fixed

- `example.js` fixed.
- Default values fixed in `README.md`.

## 1.1.3 (2018-09-18)

### Added

- `CHANGELOG.md` added.
- `eslint` module added as dev dependency
- `.eslintrc.json` file added.

### Changed

- Restructured project files.


library/index.js
// Import node modules.
const processSpawn = require('child_process').spawn;

class AudioRecorder extends require('events').EventEmitter {
  /**
   * Constructor of AudioRecord class.
   * @param {Object} options Object with optional options variables
   * @param {Object} logger Object with log, warn, and error functions
   * @returns {AudioRecorder} this
   */
  constructor(options = {}, logger) {
    super();

    // For the `rec` and `sox` only options the default is applied if a more general option is not specified.
    this._options = Object.assign({
      program: 'rec',             // Which program to use, either `arecord`, `rec`, and `sox`.
      device: null,               // Recording device to use.
      driver: null,               // Recording driver to use.

      bits: 16,					          // Sample size. (only for `rec` and `sox`)
      channels: 1,				        // Channel count.
      encoding: 'signed-integer',	// Encoding type. (only for `rec` and `sox`)
      format: 'S16_LE',           // Format type. (only for `arecord`)
      rate: 16000,                // Sample rate.
      type: 'wav',			          // File type.

      // Following options only available when using `rec` or `sox`.
      silence: 2,                 // Duration of silence in seconds before it stops recording.
      thresholdStart: 0.5,        // Silence threshold to start recording.
      thresholdStop: 0.5,         // Silence threshold to stop recording.
      keepSilence: true           // Keep the silence in the recording.
    }, options);

    this._logger = logger;

    this._childProcess = null;

    this._command = {
      arguments: [
        // Show no progress
        '-q',
        // Channel count
        '-c', this._options.channels.toString(),
        // Sample rate
        '-r', this._options.rate.toString(),
        // Format type
        '-t', this._options.type
      ],
      options: {
        encoding: 'binary',
        env: Object.assign({}, process.env)
      }
    };

    switch (this._options.program) {
      default:
      case 'sox':
        this._command.arguments.unshift(
          '-d'
        );
      case 'rec':
        // Add sample size and encoding type.
        this._command.arguments.push(
          // Show no error messages
          //   Use the `close` event to listen for an exit code.
          '-V0',
          // Endian
          //   -L = little
          //   -B = big
          //   -X = swap
          '-L',
          // Bit rate
          '-b', this._options.bits.toString(),
          // Encoding type
          '-e', this._options.encoding,
          // Pipe
          '-'
        );

        if (this._options.silence) {
          this._command.arguments.push(
            // Effect
            'silence'
          );

          // Keep the silence of the recording.
          if (this._options.keepSilence) {
            this._command.arguments.push(
              // Keep silence in results
              '-l'
            );
          }

          // Stop recording after duration has passed below threshold.
          this._command.arguments.push(
            // Enable above-periods
            '1',
            // Duration
            '0.1',
            // Starting threshold
            this._options.thresholdStart.toFixed(1).concat('%'),
            // Enable below-periods
            '1',
            // Duration
            this._options.silence.toFixed(1),
            // Stopping threshold
            this._options.thresholdStop.toFixed(1).concat('%')
          );
        }

        // Setup environment variables.
        if (this._options.device) {
          this._command.options.env.AUDIODEV = this._options.device;
        }
        if (this._options.driver) {
          this._command.options.env.AUDIODRIVER = this._options.driver;
        }
        break;

      case 'arecord':
        if (this._options.device) {
          this._command.arguments.unshift('-D', this._options.device);
        }
        this._command.arguments.push(
          // Format type
          '-f', 'S16_LE'
        );
        break;
    }

    if (this._logger) {
      // Log command.
      this._logger.log(`AudioRecorder: Command '${this._options.program} ${this._command.arguments.join(' ')}'; Options: AUDIODEV ${this._command.options.env.AUDIODEV ? this._command.options.env.AUDIODEV : '(default)'}, AUDIODRIVER: ${this._command.options.env.AUDIODRIVER ? this._command.options.env.AUDIODRIVER : '(default)'};`);
    }

    return this;
  }
  /**
   * Creates and starts the audio recording process.
   * @returns {AudioRecorder} this
   */
  start () {
    if (this._childProcess) {
      if (this._logger) {
        this._logger.warn('AudioRecorder: Process already active, killed old one started new process.');
      }
      this._childProcess.kill();
    }

    // Create new child process and give the recording commands.
    this._childProcess = processSpawn(this._options.program, this._command.arguments, this._command.options);

    // Store this in `self` so it can be accessed in the callback.
    let self = this;
    this._childProcess.on('close', function (exitCode) {
      if (self._logger) {
        self._logger.log(`AudioRecorder: Exit code '${exitCode}'.`);
      }
      self.emit('close', exitCode);
    });
    this._childProcess.on('error', function (error) {
      self.emit('error', error);
    });
    this._childProcess.on('end', function () {
      self.emit('end');
    });

    if (this._logger) {
      this._logger.log('AudioRecorder: Started recording.');
    }

    return this;
  }
  /**
   * Stops and removes the audio recording process.
   * @returns {AudioRecorder} this
   */
  stop () {
    if (!this._childProcess) {
      if (this._logger) {
        this._logger.warn('AudioRecorder: Unable to stop recording, no process active.');
      }
      return this;
    }

    this._childProcess.kill();
    this._childProcess = null;

    if (this._logger) {
      this._logger.log('AudioRecorder: Stopped recording.');
    }

    return this;
  }
  /**
   * Get the audio stream of the recording process.
   * @returns {Readable} The stream.
   */
  stream () {
    if (!this._childProcess) {
      if (this._logger) {
        this._logger.warn('AudioRecorder: Unable to retrieve stream, because no process not active. Call the start or resume function first.');
      }
      return null;
    }

    return this._childProcess.stdout;
  }
}

module.exports = AudioRecorder;
